marlowe-dashboard-client/src/Play/State.purs

```diff
@@ -95,93 +93,42 @@ handleAction CloseCard = do
-    toContractHome $ ContractHome.handleAction $ ContractHome.OpenContract contractId
-
--- other template actions
-handleAction (TemplateAction templateAction) = toTemplate $ Template.handleAction templateAction
-
--- contract home actions that need to be handled here
-handleAction (ContractHomeAction (ContractHome.OpenTemplateLibraryCard)) = handleAction $ OpenCard TemplateLibraryCard
-
-handleAction (ContractHomeAction a@(ContractHome.OpenContract _)) = do
-  toContractHome $ ContractHome.handleAction a
-  handleAction $ OpenCard ContractCard
-
--- other contract home actions
-handleAction (ContractHomeAction contractAction) = void $ toContractHome $ ContractHome.handleAction contractAction
-
--- contract actions that need to be handled here
-handleAction (ContractAction (Contract.AskConfirmation action)) = handleAction $ OpenCard $ ContractActionConfirmationCard action
-
-handleAction (ContractAction (Contract.ConfirmAction action)) = do
-  void $ toContract $ Contract.handleAction $ Contract.ConfirmAction action
-  handleAction CloseCard
-
-handleAction (ContractAction Contract.CancelConfirmation) = handleAction CloseCard
-
--- other contract  actions
-handleAction (ContractAction contractAction) = void $ toContract $ Contract.handleAction contractAction
-
--- all other actions are handled in `MainFrame.State`
-handleAction _ = pure unit
+handleAction (TemplateAction templateAction) = case templateAction of
+  Template.SetTemplate template -> do
+    mCurrentTemplate <- peruse (_templateState <<< _template)
+    when (mCurrentTemplate /= Just template) $ assign _templateState $ Template.mkInitialState template
+    handleAction $ SetScreen TemplateScreen
+  Template.OpenTemplateLibraryCard -> handleAction $ OpenCard TemplateLibraryCard
+  Template.OpenCreateWalletCard tokenName -> handleAction $ OpenCard $ CreateWalletCard $ Just tokenName
+  Template.OpenSetupConfirmationCard -> handleAction $ OpenCard ContractSetupConfirmationCard
+  Template.CloseSetupConfirmationCard -> do
+    cards <- use _cards
+    case lastOf traversed cards of
+      Just ContractSetupConfirmationCard -> handleAction CloseCard
+      _ -> pure unit
+  _ -> toTemplate $ Template.handleAction templateAction
+
+handleAction (ContractHomeAction contractHomeAction) = case contractHomeAction of
+  ContractHome.OpenTemplateLibraryCard -> handleAction $ OpenCard TemplateLibraryCard
+  a@(ContractHome.OpenContract _) -> do
+    toContractHome $ ContractHome.handleAction a
+    handleAction $ OpenCard ContractCard
+  _ -> toContractHome $ ContractHome.handleAction contractHomeAction
+
+handleAction (ContractAction contractAction) = case contractAction of
+  Contract.AskConfirmation action -> handleAction $ OpenCard $ ContractActionConfirmationCard action
+  Contract.ConfirmAction action -> do
+    void $ toContract $ Contract.handleAction $ Contract.ConfirmAction action
+    handleAction CloseCard -- TODO: guard against closing the wrong card
```

Comment:
I wonder if it makes sense to add as a parameter to `CloseCard` which card are we trying to close and make the guard work in there instead of on the calling site

Reply:
🤔 yeah quite possibly, I'll try that